sphinx/ext/autodoc/__init__.py

```diff
@@ -1485,13 +1494,32 @@ def add_directive_header(self, sig: str) -> None:
         if not self.doc_as_attr and self.options.show_inheritance:
             sourcename = self.get_sourcename()
             self.add_line('', sourcename)
-            if hasattr(self.object, '__bases__') and len(self.object.__bases__):
-                bases = [':class:`%s`' % b.__name__
-                         if b.__module__ in ('__builtin__', 'builtins')
-                         else ':class:`%s.%s`' % (b.__module__, b.__qualname__)
-                         for b in self.object.__bases__]
-                self.add_line('   ' + _('Bases: %s') % ', '.join(bases),
-                              sourcename)
+
+            def restify(cls: "Type") -> str:
+                """Convert class to reST text."""
+                if cls.__module__ in ('__builtin__', 'builtins'):
+                    return ':class:`%s`' % cls.__name__
+                elif inspect.isgenericalias(cls):
+                    if cls._name:
+                        text = ':class:`%s.%s`' % (cls.__module__, cls._name)
+                    else:
+                        text = restify(cls.__origin__)
+
+                    if cls.__args__:
+                        text += r"\ [%s]" % ", ".join(restify(a) for a in cls.__args__)
+                    return text
+                elif hasattr(cls, '__qualname__'):
+                    return ':class:`%s.%s`' % (cls.__module__, cls.__qualname__)
+                else:
+                    # not a class (ex. TypeVar)
+                    return ':obj:`%s.%s`' % (cls.__module__, cls.__name__)
+
+            if hasattr(self.object, '__orig_bases__') and len(self.object.__orig_bases__):
+                bases = [restify(cls) for cls in self.object.__orig_bases__]
+                self.add_line('   ' + _('Bases: %s') % ', '.join(bases), sourcename)
+            elif hasattr(self.object, '__bases__') and len(self.object.__bases__):
```

Comment:
Is this still needed?

Reply:
Yes, non-generic-subclasses don't have `__orig_bases__` even after PEP-560. So this code is still needed.